### PR TITLE
Add HttpTransport extensibility and synchronous serialization support

### PR DESCRIPTION
**Pull request by mattjohnsonpint**
As mentioned in https://github.com/getsentry/sentry-dotnet/pull/1553#issuecomment-1081302669, we need to have a clean way for higher-level SDKs (like sentry-unity) to implement their own HTTP transports.  In doing so, we'll need to support both asynchronous and synchronous handling of response messages, including synchronous serialization support for things like logging of payloads.

This PR takes most of the functionality of `HttpTransport` and pulls it into an abstract base class, `HttpTransportBase`.  That allows protected access to the core functionality of the transport, reducing `SendEnvelopeAsync` to a small bit of orchestration code.  Higher-level SDKs (such as `UnityWebRequestTransport` proposed in getsentry/sentry-unity#657) can implement their own version of that method without having to take a dependency on `HttpClient` or async/await.

Note that `ITransport` is still internal and implemented only on `HttpTransport` - not `HttpTransportBase`.

A fully synchronous path for handling response messages is available as `HandleResponse` instead of `HandleResponseAsync` (either can be used by the parent class).  In order to fully implement this correctly, `Sentry.Protocol.Envelopes.ISerializable` has been given a `Serialize` method which mirrors the existing `SerializeAsync` method.  All implementations and tests are identical, except as to whether async methods are employed or not.  At the lowest level, these now use things like `Stream.Write` instead of `Stream.WriteAsync` - which are implemented in .NET using the correct primitives (no sync over async or vice versa).

### CODE REVIEW

**Comment by mattjohnsonpint** [comment]
Just because it's a bit hard to read on GitHub, here's the new version of `HttpTransport.SendEnvelopeAsync` in its entirety.  This is now the only method in the class.

```csharp
public async Task SendEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken = default)
{
    using var processedEnvelope = ProcessEnvelope(envelope);
    if (processedEnvelope.Items.Count > 0)
    {
        using var request = CreateRequest(processedEnvelope);
        using var response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        await HandleResponseAsync(response, processedEnvelope, cancellationToken).ConfigureAwait(false);
    }
}
```

**Review by vaind — APPROVED** [review]
Works fine for the unity WebGL implementation. There's just one open point that could maybe be improved. See https://github.com/getsentry/sentry-unity/pull/657#pullrequestreview-927328402

**Review by bruno-garcia — APPROVED** [review]
LGTM

File: src/Sentry/Envelopes/Envelope.cs [hunk 1]
```diff
@@ -62,6 +62,13 @@ await using (writer.ConfigureAwait(false))
             }
         }
 
+        private void SerializeHeader(Stream stream, IDiagnosticLogger? logger)
+        {
+            using var writer = new Utf8JsonWriter(stream);
```
[the remark this answers is not in the document]
Comment (mattjohnsonpint) [reply src/Sentry/Envelopes/Envelope.cs:67]: That's a concern with `StreamWriter`/`StreamReader`, but not for `Utf8JsonWriter`.  Its dispose just flushes the stream, it doesn't dispose it.

**Comment by mattjohnsonpint** [comment]
I removed the `HttpContentReader`,  and just added a synchronous serialization path for `EnvelopeHttpContent`.

**Pull request merged**